webrender_api/src/display_item.rs

```diff
@@ -89,6 +89,58 @@ impl LayerPrimitiveInfo {
             tag: None,
         }
     }
+
+    pub fn decompose(
+        &self,
+        tile_size: LayerSize,
+        tile_spacing: LayerSize,
+        max_prims: usize,
+    ) -> Vec<LayerPrimitiveInfo> {
+        let mut prims = Vec::new();
+        let tile_repeat = tile_size + tile_spacing;
+
+        if tile_repeat.width <= 0.0 ||
+           tile_repeat.height <= 0.0 {
+            return prims;
+        }
+
+        if tile_repeat.width < self.rect.size.width ||
+           tile_repeat.height < self.rect.size.height {
+            let local_clip = self.local_clip.clip_by(&self.rect);
+            let rect_p0 = self.rect.origin;
+            let rect_p1 = self.rect.origin + self.rect.size;
+
+            let mut y0 = rect_p0.y;
+            while y0 < rect_p1.y {
+                let mut x0 = rect_p0.x;
+
+                while x0 < rect_p1.x {
+                    prims.push(PrimitiveInfo {
```

Comment:
Should we skip emitting primitives that are completely clipped out by `local_clip` or is this already taken care of elsewhere?

Reply:
They are handled elsewhere - we could potentially save a bit of memory allocation by doing it early here though.